### PR TITLE
Point selection in 3d

### PR DESCRIPTION
**Pull request by kevinyamauchi**
# Description
In this PR, @alisterburt and I introduce the functionality to select individual points in 3D. We will add the ability to perform selection of "rectangles of points" (i.e., click and drag) in a follow up PR.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by kevinyamauchi** [comment]
I think we have a version of this working! We can now select points in 3D and drag them around. We can also add points in clicking, but the behavior isn't intuitive unless one is familiar with how the camera works.

https://user-images.githubusercontent.com/1120672/140504905-6f15da48-25d7-4286-a0b4-0e6a72cf950f.mov

A couple of notes:
- Selection works via clicking, but we have not implemented the click+drag lasso yet. One can select multiple points via shift+click.
- Once a point is selected, one can do all of the expected stuff (move, change display properties, delete)
- Adding points in 3D works. It puts the added point on the plane defined by the camera center and the view direction. Eventually, we would like to add a plane overlay that shows the plane the points will be added on. That requires a lot more (e.g., a proper plane overlay system), so it will have to wait for a future PR.

Related to adding points in 3D:  we aren't sure if this should be on the layer control modes since it isn't obvious where the plane the points are added on lays within the scene. However, the current machinery for disabling modes (i.e., layer.editable) doesn't readily offer the fine grained control for disabling one "editing" mode (just `add` in this case). One possibility that would make it easy is to add a layer.ndisplay event, but we understand there is a possibility we will want to remove dims info from the layer (although @alisterburt and I are in favor of putting dims on the layer and removing the data). Thoughts?

**Comment by jni** [comment]
So cool!!! :heart_eyes: 

Yeah it took me a few seconds even after your explanation to understand where the points were landing. One way to think of it, though, is that they are landing exactly on the visible plane if one turns on @alisterburt's "slicing plane follows camera" mode, *without* needing to refer to the other layer, and I think that's pretty brilliant! So I'm actually in favour of leaving it, it will only be confusing for a short while, and it's not like it's easy to get into add mode "accidentally". That way we can punt on the architectural question which is pretty hairy. :stuck_out_tongue_winking_eye:

**Comment by alisterburt** [comment]
Thanks @jni ! Agree that this is definitely a fun interim solution - the fact that moving nD points already works is dreamy too 😍

In that case, let's push on!

@kevinyamauchi I'll try to add a guard so that click and drag doesn't make anything fall over this evening

File: napari/layers/points/_points_mouse_bindings.py
Comment (kevinyamauchi) [inline napari/layers/points/_points_mouse_bindings.py:61]: @alisterburt , I added this guard to prevent errors when clicking + dragging in select mode. I _think_ it works. Was there another guard you had in mind for click+drag?

**Comment by kevinyamauchi** [comment]
> So I'm actually in favour of leaving it, it will only be confusing for a short while, and it's not like it's easy to get into add mode "accidentally". That way we can punt on the architectural question which is pretty hairy.

I agree, @jni .

@alisterburt , I think this PR is pretty much done. I added tests for the `get_value()` stuff we added. From my side, I think the main thing to add are tests for the mouse callbacks. What do you think?

File: napari/layers/points/_points_mouse_bindings.py
Comment (alisterburt) [reply napari/layers/points/_points_mouse_bindings.py:61]: oh, perfect! Nope that's all we need, I was just out of date 👍🏼

File: napari/layers/points/_points_mouse_bindings.py
Comment (alisterburt) [inline napari/layers/points/_points_mouse_bindings.py]:
```suggestion
            # only allow click+drag bounding box in 2D
```

**Comment by alisterburt** [comment]
@kevinyamauchi I actually have a bit of time now so will try to get that test in!

**Review by sofroniewn — APPROVED** [review]
LGTM! I havn't given this a try, but movie looks great and I'm happy with the code changes - I'd say feel free to merge when ready!!

File: napari/layers/points/points.py
Comment (brisvag) [inline napari/layers/points/points.py:1336]: Same comment I had elsewhere: this method name is a bit confusing to me, since it sounds like a verb ("display this!"). Same is true of the base layer method!

File: napari/layers/points/points.py
Comment (brisvag) [inline napari/layers/points/points.py:1364]: Just a heads up that we might want to think about: in #3430 (which will be merged only after #3494 is done) the points layer will have a new option, which is to consider the size in screen pixels, rather than data pixels. This will probably break, then.

**Review by brisvag — COMMENTED** [review]
Great stuff, looking forward to using this! I have a couple of minor comments, nothing blocking there.

As for the 3D picking: I think we should be careful with enabling something like this by default, because I think most of the use cases for picking points in 3D will be relying on an underlying geometry to determine the depth of the point (isosurfaces, mip, slicing/clipping planes, labels...). I struggle to see a use case where one wants to move the camera in order to decide where to put a point. The only exception is the example of the slice-following-the-camera by @alisterburt , but even then it's just one of the cases above.

File: napari/layers/points/points.py
Comment (kevinyamauchi) [reply napari/layers/points/points.py:1364]: Good to know! Is there a plan for a function that allows for conversion between size in canvas units and size in data units? 

Since we are using the same code as the 2D layer for determining which points are selected, I think it is okay to proceed with this for now, as whatever is done in #3430 to make `Points.get_value()` work in 2D should also make 3D work.

**Comment by kevinyamauchi** [comment]
> As for the 3D picking: I think we should be careful with enabling something like this by default, because I think most of the use cases for picking points in 3D will be relying on an underlying geometry to determine the depth of the point (isosurfaces, mip, slicing/clipping planes, labels...). I struggle to see a use case where one wants to move the camera in order to decide where to put a point. The only exception is the example of the slice-following-the-camera by @alisterburt , but even then it's just one of the cases above.

Hey @brisvag , I agree that the current point adding behavior is a bit unintuitive. For the reasons @jni gave above (https://github.com/napari/napari/pull/3508#issuecomment-962286706), I think it is okay to move forward with it exposed.

File: napari/layers/points/points.py
Comment (kevinyamauchi) [reply napari/layers/points/points.py:1336]: I can see how that would be confusing. Do you have a name suggestion? just `bounding_box_augmented`?

File: napari/layers/base/base.py
Comment (alisterburt) [inline napari/layers/base/base.py]:
```suggestion 
       """Calculate a (point, normal) plane parallel to the canvas in data
        coordinates, centered on the centre of rotation of the camera.
```

File: napari/layers/base/base.py
Comment (alisterburt) [inline napari/layers/base/base.py:1261]: we should add to the 3d interactivity docs a description of things down the road, definitely not blocking merge for this!

File: napari/layers/points/points.py
Comment (alisterburt) [reply napari/layers/points/points.py:1336]: I think `displayed_bounding_box` or `scene_bounding_box` would be good BUT this method is mirroring the (not well named) property on the base layer.

In general we don't have a consistent way of describing the 'displayed_world' coordinate system, usually called `scene` in other programs, and should think about this a bit but I think for the purposes of this PR this name is correct and we should refactor separately. Thoughts?

**Review by alisterburt — APPROVED** [review]
Increased test coverage looks great @kevinyamauchi !

I left a couple of comments but I don't think there's anything blocking here - we should make a separate issue for discussing the naming conventions around the 'displayed world' coordinate system and unify in one fell swoop once we have consensus there.  I think we should keep `display_bounding_box_augmented` as is because the underlying method is still called `display_bounding_box` - @brisvag does that sound reasonable?

I'm approving! 🙂 Great stuff

File: napari/layers/base/base.py
Comment (kevinyamauchi) [reply napari/layers/base/base.py:1261]: Agreed!

File: napari/layers/points/points.py
Comment (brisvag) [reply napari/layers/points/points.py:1336]: Yeah, sounds good to me. I do think both this and the base layer one should be renamed, but we can keep it for another PR :)

File: napari/layers/points/points.py
Comment (brisvag) [reply napari/layers/points/points.py:1364]: Yup! this should definitely be addressed over there with some function like that. I just wanted to set a reminder for myself!

File: napari/layers/base/base.py
Comment (brisvag) [inline napari/layers/base/base.py:1283]: This says that the click position is in "world coordinates". Could you elaborate what that means? If it's just any position in world space, why is it specifically "click position"? Is this more generalizable than it appears?

File: napari/layers/utils/interactivity_utils.py
Comment (brisvag) [inline napari/layers/utils/interactivity_utils.py]: Why at the near side?

File: napari/utils/geometry.py
Comment (brisvag) [inline napari/utils/geometry.py]: Can you elaborate this a bit? Is this whole plane getting rotated? Are they being projected onto the rotated plane? What's the difference of this compared to simply applying a rotation matrix to a bunch of non-coplanar points?

**Review by brisvag — DISMISSED** [review]
In terms of functionality, everything seems to be working! Great PR :)

Reading the conversation and the doscstrings, I'm still unsure *where* exactly things happen. It would be nice to add in the docstrings some more details to make sure that future changes don't break anything, since the geometry here is quite hard and I'm definitely not following what happens just by reading the code.

One thing that confuses me specifically is: where does the "plane through the center of rotation of the camera" happen? In `click_plane_from_click_data`? I added a few individual comments about this!

File: napari/layers/base/base.py
Comment (alisterburt) [reply napari/layers/base/base.py:1283]: I've had trouble with this too - the key is that `event.position` from the mouse event is a position in nD world coordinates. That position is on a plane, centered on the `Camera.center`, parallel to the canvas. This method transforms this plane (defined by the click position and view direction) into data coordinates so you can use it to query things in the data coordinate system. Does that help?

Comment (brisvag) [reply napari/layers/base/base.py:1283]: Yup, that helps! I propose to add some of this to the docstring, so it's clear what "click_position" represents: not just *any* position, but one that's on a plane passing through `Camera.center` and parallel to the canvas.

**Comment by kevinyamauchi** [comment]
Thanks for the review, @brisvag . I agree we could improve the documentation. I have some time to update the docstrings later today.

File: napari/layers/base/base.py
Comment (alisterburt) [reply napari/layers/base/base.py:1261]: after discussion, let's rename this to `_world_to_data_plane`

File: napari/utils/geometry.py
Comment (alisterburt) [reply napari/utils/geometry.py]: @brisvag agreed this isn't super clear, the transformation applied on the points is one which aligns the two plane normals - we're renaming

Comment (alisterburt) [reply napari/utils/geometry.py]: naming is hard 😆

**Comment by sofroniewn** [comment]
@brisvag can you give this a review. If it looks good to you then I think we can merge. Thanks!!

**Comment by sofroniewn** [comment]
@brisvag i believe your requested changes have now been addressed. I will merge now, but do let us know if you'd still like to see more things. Thanks for the review!!

**Comment by brisvag** [comment]
Sorry for the delay, thanks for moving it forward! Yes, everything's clearer now :)

**Pull request merged**
